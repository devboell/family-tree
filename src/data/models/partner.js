const knex = require('../connectors')
const sql = require('./sql')


const findByPersonId = personId =>
  knex.raw(sql.partners, personId)

const unassignChildren = relationIds =>
  knex('Person')
    .whereIn('bornToId', relationIds)
    .update('bornToId', null)

const findChildren = relationIds =>
  knex('Person').whereIn('bornToId', relationIds)


module.exports = {
  findByPersonId,
  unassignChildren,
  findChildren,
}
