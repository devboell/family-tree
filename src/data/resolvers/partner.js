const omit = require('lodash/fp/omit')
const difference = require('lodash/fp/difference')
const PartnerModel = require('../models/partner')
const RelationModel = require('../models/relation')


const assignChildrenToPartner = (partners, allChildren) =>
  partners.map(partnerWithId => ({
    relationId: partnerWithId.relationId,
    partner: omit('relationId', partnerWithId),
    children: allChildren.filter(child => child.bornToId === partnerWithId.relationId),
  }))

const partnersForPerson = async (person) => {
  const partners = await PartnerModel.findByPersonId(person.id)
  const relationIds = partners.map(partner => partner.relationId)
  const children = await PartnerModel.findChildren(relationIds)

  const partnersWithChildren = assignChildrenToPartner(partners, children)
  return partnersWithChildren
}

const partnerIds = async (person) => {
  const partners = await PartnerModel.findByPersonId(person.id)
  return partners.map(partner => partner.id)
}

const createPartners = async (personId, partnerInputs) =>
  RelationModel.create(personId, partnerInputs.map(partner => partner.partnerId))

const updatePartners = async (personId, partnerInputs) => {
  const relations = await RelationModel.findByPersonId(personId)
  const relationIds = relations.map(relation => relation.id)
  const relationsToDelete = difference(relationIds)(partnerInputs
    .map(prt => prt.relationId))

  if (relationsToDelete.length) {
    await RelationModel.remove(relationsToDelete)
    await PartnerModel.unassignChildren(relationsToDelete)
  }
  const newPartnerIds = partnerInputs
    .filter(prt => prt.relationId === null)
    .map(prt => prt.partnerId)

  if (newPartnerIds.length) {
    await RelationModel.create(personId, newPartnerIds)
  }
}

const removePartners = async (personId) => {
  const relations = await RelationModel.findByPersonId(personId)
  const relationIds = relations.map(relation => relation.id)

  if (relationIds.length) {
    await RelationModel.remove(relationIds)
    await PartnerModel.unassignChildren(relationIds)
  }
}

module.exports = {
  partnersForPerson,
  partnerIds,
  createPartners,
  updatePartners,
  removePartners,
}
