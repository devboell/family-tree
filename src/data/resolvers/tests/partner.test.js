import knex from '../../connectors'
import partner from '../partner'

describe('partner, resolvers', () => {
  beforeEach(async () => {
    await knex.migrate.rollback()
    await knex.migrate.latest()
    await knex.seed.run()
  })

  afterEach(async () => {
    await knex.migrate.rollback()
  })

  afterAll(() => {
    knex.destroy()
  })

  it('partnersForPerson', async () => {
    const person = { id: 4 }
    const partners = await partner.partnersForPerson(person)
    expect(partners).toMatchSnapshot()
  })

  it('partnerIds', async () => {
    const person = { id: 4 }
    const partnerIds = await partner.partnerIds(person)
    expect(partnerIds).toMatchSnapshot()
  })

  it('createPartners', async () => {
    const partnerInputs = [{ partnerId: 6 }, { partnerId: 7 }]
    await partner.createPartners(4, partnerInputs)
    const partnerIds = await partner.partnerIds({ id: 4 })

    expect(partnerIds).toMatchSnapshot()
  })

  it('updatePartners', async () => {
    const partnerInputs = [
      { partnerId: 14, relationId: 5 },
      { partnerId: 7, relationId: null },
    ]
    await partner.updatePartners(4, partnerInputs)
    const partnerIds = await partner.partnerIds({ id: 4 })

    expect(partnerIds).toMatchSnapshot()
  })
})
