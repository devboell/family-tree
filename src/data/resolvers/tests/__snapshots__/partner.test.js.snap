// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`partner, resolvers createPartners 1`] = `
Array [
  5,
  14,
  6,
  7,
]
`;

exports[`partner, resolvers partnerIds 1`] = `
Array [
  5,
  14,
]
`;

exports[`partner, resolvers partnersForPerson 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "birthDate": 1103,
        "bornToId": 2,
        "deathDate": 1120,
        "gender": "M",
        "id": 6,
        "isMonarch": 0,
        "name": "William Adelin",
        "title": "Duke of Normandy",
      },
      Object {
        "birthDate": 1102,
        "bornToId": 2,
        "deathDate": 1167,
        "gender": "F",
        "id": 7,
        "isMonarch": 0,
        "name": "Empress Matilde",
        "title": "Holy Roman Empress",
      },
    ],
    "partner": Object {
      "birthDate": 1080,
      "bornToId": null,
      "deathDate": 1118,
      "gender": "F",
      "id": 5,
      "isMonarch": 0,
      "name": "Matilde of Scotland",
      "title": "Queen of England",
    },
    "relationId": 2,
  },
  Object {
    "children": Array [],
    "partner": Object {
      "birthDate": 1103,
      "bornToId": null,
      "deathDate": 1151,
      "gender": "F",
      "id": 14,
      "isMonarch": 0,
      "name": "Adeliza of Louvain",
      "title": "Queen of England",
    },
    "relationId": 5,
  },
]
`;

exports[`partner, resolvers updatePartners, no changes 1`] = `
Array [
  5,
  14,
]
`;

exports[`partner, resolvers updatePartners, one partner added 1`] = `
Array [
  5,
  14,
  7,
]
`;

exports[`partner, resolvers updatePartners, one partner deleted 1`] = `
Array [
  14,
]
`;

exports[`partner, resolvers updatePartners, one partner deleted and one added 1`] = `
Array [
  14,
  7,
]
`;
